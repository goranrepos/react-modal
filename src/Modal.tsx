import React, { useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import "./Modal.css";

interface IModal {
    hide: () => void;
}

const Modal = ({ hide }:IModal) => {

    const modalRef = useRef<HTMLDivElement>(null);
    const handleTabKey = (e: KeyboardEvent): void => {
        if(!modalRef.current) return;
      const focusableModalElements = modalRef.current.querySelectorAll(
        'a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select'
      );
      const firstElement = focusableModalElements[0];
      const lastElement =
        focusableModalElements[focusableModalElements.length - 1];
  
      if (!e.shiftKey && document.activeElement === lastElement) {
        (firstElement as HTMLElement).focus();
       
        return e.preventDefault();
      }
  
      if (e.shiftKey && document.activeElement === firstElement) {
         (lastElement as HTMLElement).focus();
         
        e.preventDefault();
      }
    };

    const keyListenersMap =[
        {
            key: 'Escape',
            value: hide
        }, {
            key: 'Tab',
            value: handleTabKey 
        }];

    useEffect(() => {
        function keyListener(e: KeyboardEvent) {
            
          const listener = keyListenersMap.find(keyListener => {     
            return (keyListener.key === e.key) })?.value;
          return listener && listener(e);
        }
        document.addEventListener("keydown", keyListener);
    
        return () => document.removeEventListener("keydown", keyListener);
      });

   return (
    ReactDOM.createPortal(
        <React.Fragment>
          <div className="modal-overlay"/>
          <div className="modal-wrapper" aria-modal role="dialog" aria-labelledby="dialog1_label">
            <div className="modal" ref={modalRef}>
              <div className="modal-header">
                <button type="button" data-testid="closeMainModal" className="modal-close-button" data-dismiss="modal" aria-label="Close" onClick={hide}>
                  <span aria-hidden="true">&times;</span>
                </button>
              </div>
              <div className="modal-body">
              <h2 id="dialog1_label" className="dialog_label" tabIndex={-1}>Modal Title</h2>
              <p>
                Hello, I'm a modal.
              </p>
              </div>
             
              <div className="modal-footer">
                <button data-testid="continueModal">Continue</button>
                <button
                  onClick={hide}
                  id="cancelBtn"
                  data-testid="cancelButton"
                >
                  Cancel
                </button>
                
              </div>
            </div>
          </div>
        </React.Fragment>, document.body
      )
   );
    };

export default Modal;