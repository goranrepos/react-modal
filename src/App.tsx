// import React from 'react';
// import logo from './logo.svg';
// import './App.css';

// function App() {
//   return (
//     <div className="App">
//       <header className="App-header">
//         <img src={logo} className="App-logo" alt="logo" />
//         <p>
//           Edit <code>src/App.tsx</code> and save to reload.
//         </p>
//         <a
//           className="App-link"
//           href="https://reactjs.org"
//           target="_blank"
//           rel="noopener noreferrer"
//         >
//           Learn React
//         </a>
//       </header>
//     </div>
//   );
// }

// export default App;

import React from 'react';
import { useRef } from "react";
import Modal from "./Modal";
import "./styles.css";
import useModal from "./useModal";

/*
The modal is following the Dialog (Modal) Pattern
as defined in https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/

Guidelines that were specifically implemented:
1. users cannot interact with content outside an active dialog window
2. Inert content outside an active dialog is typically visually obscured
3. Tab and Shift + Tab do not move focus outside the dialog. Moves focus to the next tabbable element inside the dialog.
4. When a dialog opens, focus moves to an element contained in the dialog
5. The modal opens with focus on the title, enables screen readers to announce the description along with the dialog title and initially focused element when the dialog opens, which is typically helpful only when the descriptive content is simple and can easily be understood without structural information.
6. Dialog container element has aria-modal set to true.
7. The element that serves as the dialog container has a role of dialog.
8. A value set for the aria-labelledby property that refers to a visible dialog title.
9. Escape: Closes the dialog.
10. When a dialog closes, focus returns to the element that invoked the dialog.
*/

export default function App() {
  const {isShowing, toggle } = useModal();

  const modalButtonRef = useRef<HTMLButtonElement>(null);

 const handleHide = () => {
    modalButtonRef.current?.focus();
    toggle(); 
  }

  return (
    <div className="App">
      <h1 data-testid="appTitle">AppTitle</h1>
      <h2>Letâ€™s see a modal</h2>
      <button  type="button" className="button-default" onClick={toggle} ref={modalButtonRef} name="openModal">Show Modal</button>
      {isShowing && <Modal
        hide={handleHide}
        data-testid="mainModal"
      />}
    </div>
  );
}
