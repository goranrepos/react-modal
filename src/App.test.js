import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";
import Modal from "./Modal";

describe("Application with Modal", () => {
  // cleanup after each test
  afterEach(cleanup);

  test("render h1 heading of the application", () => {
    render(<App />);
    const elementH1 = screen.queryByTestId(/appTitle/i);
    expect(elementH1).toBeInTheDocument();
  });

  test("render App title", () => {
    render(<App />);
    expect(screen.getByText(/AppTitle/i)).toBeInTheDocument();
  });

  test("render open modal button", () => {
    render(<App />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("render non-active modal as null by default", () => {
    render(<App />);
    const element = screen.queryByTestId(/mainModal/i);
    expect(element).toBeNull();
  });

  test("render application, simulate open modal button click that opens the modal", async () => {
    const { baseElement } = render(<App />);
    fireEvent.click(screen.queryByText("Show Modal"));
    expect(baseElement).toMatchSnapshot();
    const modal = await screen.findByText("Modal Title");
    expect(modal).toBeTruthy();
  });

  test("render modal and modal has one heading only", async () => {
    const handleHide = jest.fn;
    render(<Modal hide={handleHide} data-testid="mainModal" />);
    const headings = await screen.findAllByRole("heading");
    const headingLength = headings.length;
    expect(headingLength).toBe(1);
  });

  test("render modal and modal has three buttons", async () => {
    const handleHide = jest.fn;
    render(<Modal hide={handleHide} data-testid="mainModal" />);
    const buttons = await screen.findAllByRole("button");
    const buttonLength = buttons.length;
    expect(buttonLength).toBe(3);
  });

  test("render application, open modal, and close modal with close button click", async () => {
    render(<App />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    const closeModalButton = await screen.findByTestId(/closeMainModal/i);
    fireEvent.click(closeModalButton);
    expect(screen.queryByText("Modal Title")).toBeNull();
  });

  test("render application, open modal, and close modal with cancel button click", async () => {
    render(<App />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    const cancelButton = await screen.findByTestId(/cancelButton/i);
    fireEvent.click(cancelButton);
    expect(screen.queryByText("Modal Title")).toBeNull();
  });

  test("render application, open modal using keyboard to focus on the button and press ENTER", async () => {
    render(<App />);
    const button = screen.queryByText("Show Modal");
    button.focus();
    await userEvent.keyboard("[Enter]");
    const cancelButton = await screen.findByTestId(/cancelButton/i);
    await waitFor(() => expect(cancelButton).toBeInTheDocument());
  });

  test("render application, open modal, and close modal with focus on cancel button and press ENTER", async () => {
    render(<App />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    const cancelButton = await screen.findByTestId(/cancelButton/i);
    cancelButton.focus();
    await userEvent.keyboard("[Enter]");
    expect(screen.queryByText("Modal Title")).toBeNull();
  });

  test("render application, open modal, and close modal with focus on close button  and press ENTER", async () => {
    render(<App />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    const closeModalButton = await screen.findByTestId(/closeMainModal/i);
    closeModalButton.focus();
    await userEvent.keyboard("[Enter]");
    expect(screen.queryByText("Modal Title")).toBeNull();
  });

  test("user is locked in the modal and can tab through buttons using keyboard", async () => {
    const handleHide = jest.fn;
    render(<Modal hide={handleHide} data-testid="mainModal" />);
    const closeButton = screen.getByTestId(/closeMainModal/i);
    const continueModal = screen.getByTestId(/continueModal/i);
    const cancelButton = screen.getByTestId(/cancelButton/i);
    await userEvent.tab();
    expect(closeButton).toHaveFocus();
    await userEvent.tab();
    expect(continueModal).toHaveFocus();
    await userEvent.tab();
    expect(cancelButton).toHaveFocus();
    await userEvent.tab();
    expect(closeButton).toHaveFocus();
    fireEvent.keyDown(closeButton, {
      key: "Tab",
      code: "Tab",
      charCode: 9,
      shiftKey: true,
    });
    expect(cancelButton).toHaveFocus();
  });
});
