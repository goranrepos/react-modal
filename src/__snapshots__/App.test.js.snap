// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application with Modal render application, simulate open modal button click that opens the modal 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <h1
        data-testid="appTitle"
      >
        AppTitle
      </h1>
      <h2>
        Let’s see a modal
      </h2>
      <button
        class="button-default"
        name="openModal"
        type="button"
      >
        Show Modal
      </button>
    </div>
  </div>
  <div
    class="modal-overlay"
  />
  <div
    aria-labelledby="dialog1_label"
    aria-modal="true"
    class="modal-wrapper"
    role="dialog"
  >
    <div
      class="modal"
    >
      <div
        class="modal-header"
      >
        <button
          aria-label="Close"
          class="modal-close-button"
          data-dismiss="modal"
          data-testid="closeMainModal"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            ×
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <h2
          class="dialog_label"
          id="dialog1_label"
          tabindex="-1"
        >
          Modal Title
        </h2>
        <p>
          Hello, I'm a modal.
        </p>
      </div>
      <div
        class="modal-footer"
      >
        <button
          data-testid="continueModal"
        >
          Continue
        </button>
        <button
          data-testid="cancelButton"
          id="cancelBtn"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</body>
`;
